mirrord/agent/src/sniffer.rs

```diff
@@ -491,7 +508,42 @@ impl TcpConnectionSniffer {
         Ok(())
     }
 
-    #[tracing::instrument(level = "trace", skip(self, eth_packet), fields(bytes = %eth_packet.len()))]
+    /// First it checks the `tcp_flags` with [`is_new_connection`], if that's not the case, meaning
+    /// we have traffic from some existing connection from before mirrord started, then it tries to
+    /// see if `bytes` contains an HTTP request (HTTP/1) of some sort. When an HTTP request is
+    /// detected, then the agent should start mirroring as if it was a new connection.
+    #[tracing::instrument(level = "debug", ret, skip(bytes))]
+    fn treat_as_new_session(tcp_flags: u16, bytes: &[u8]) -> bool {
```

Comment:
We would want the mirror to start from the valid HTTP stream, and also a potential flow could be
Frame 1: "AAAENDOFPREVIOUSRESPONSE H"
Frame 2: "TTP"
and we wouldn't catch it, so we actually need something like we have for the steal, check current buf is http start marker - no? skip by 1 byte then try again, when bytes are exhausted:

1. if we were in middle of comparison, wait until more data is read.
2. if last byte wasn't part of comparison, you can drop the buffer.

Reply:
I was thinking the same, but couldn't make it happen with simple HTTP requests.

Like, they always come in the new packets, maybe if I send some big requests I can trigger this behavior? That's why I'm not even changing the packet bytes to start from where the regex starts matching.

Reply:
Still works when sending something like `vec![1; 1500]` as the body.

Only way I can see to trigger the invalid frame behavior is if I send this type of frame myself, as a `TcpStream` instead of sending an http request.

Reply:
I'd assume this behavior happens on HTTP proxies or high performant services where they buffer and flush to socket, that can create separate network frames. In the end the sender can't really control how the TCP frames are built. That's part of the OSI spec.

Reply:
You can test it in an integration test also (or just UT tbh)

Reply:
Yeah, I'm sure this can (and probably will) happen.

Should I:

a. go for a more complete implementation with the whole hyper and stuff right now? or;
b. merge this simple solution, as it kinda works on average cases, then work on (a)?

Reply:
Good q (and idea).. I think yes -  @eyalb181 ?

Reply:
I'm for option b

Reply:
נ